alternator - Doc - Update DescribeTable response and introduce hashing function differences

fee-mendes opened the pull request:
This commit introduces the following changes to Alternator compability doc:

* As of https://github.com/scylladb/scylladb/pull/11298 Alternator will return ProvisionedThroughput in DescribeTable API calls. We add the fact that tables will default to a BillingMode of PAY_PER_REQUEST (this wasn't made explicit anywhere in the docs), and that the values for RCUs/WCUs are hardcoded to 0.
* Mention the fact that ScyllaDB (thus Alternator) hashing function is different than AWS proprietary implementation for DynamoDB. This is mostly of an implementation aspect rather than a bug, but it may cause user confusion when/if comparing the ResultSet between DynamoDB and Alternator returned from Table Scans.
* Finally, it corrects an annoying reading typo under the Unimplemented API features section

Refs: https://github.com/scylladb/scylladb/issues/11222
Fixes: https://github.com/scylladb/scylladb/issues/11315

tzach commented on docs/alternator/compatibility.md:
```diff
@@ -16,7 +16,16 @@ provision that - Scylla requires you to provision your cluster. You need
+and `DescribeTable` API calls will return a value of 0 for the RCUs, WCUs
+and NumberOfDecreasesToday within the response.
+
+Another notable difference between DynamoDB and Scylla comes down to the
+underlying hashing algorithm. While DynamoDB uses a proprietary hashing
+function, ScyllaDB implements the well-known Cassandra Murmur3 algorithm.
```
DynamoDB and ScyllaDB use a different underlying hashing algorithm: DynamoDB uses a proprietary hashing function, while ScyllaDB implements the well-known [Murmur3](https://docs.scylladb.com/stable/glossary.html#term-Partitioner) algorithm.

psarna replied:
I agree, murmur3 is not Cassandra-specific. Although technically, ScyllaDB and all our drivers implement a buggy murmur3 flavor, which is Cassandra-specific indeed :innocent: Ref: https://github.com/scylladb/scylla-rust-driver/blob/fd06928929c0dd78a72c7aca494a998da1514a79/scylla/src/transport/partitioner.rs#L27-L29

fee-mendes replied:
Correct. Although I agree that removing "Cassandra" mentions should be a guidance in general. :-P

Fixed.

nyh commented on docs/alternator/compatibility.md:
```diff
@@ -16,7 +16,16 @@ provision that - Scylla requires you to provision your cluster. You need
+and `DescribeTable` API calls will return a value of 0 for the RCUs, WCUs
+and NumberOfDecreasesToday within the response.
+
+Another notable difference between DynamoDB and Scylla comes down to the
```
I think this paragraph doesn't belong in the "Provisioning" heading.
I hesitate if users need to know about it at all (after all, I don't think they have any guarantee that even DynamoDB won't change their hash function next week), but if they do, it should be in a different section. Maybe we can have a section heading "Hash-key hash function" or something like that?

nyh commented on docs/alternator/compatibility.md:
```diff
@@ -16,7 +16,16 @@ provision that - Scylla requires you to provision your cluster. You need
+and NumberOfDecreasesToday within the response.
+
+Another notable difference between DynamoDB and Scylla comes down to the
+underlying hashing algorithm. While DynamoDB uses a proprietary hashing
```
The paragraph introducing this difference should mention **hash key** (a.k.a. partition keys), otherwise it's not clear what this "hashing" mentioned in this paragraph refers to (a reader might guess, but I think it won't be obvious to everyone).

fee-mendes commented:
Changes:

- Added link to Murmu3, removed references to Cassandra
- Added a new "Scan ordering" heading
- Introduced the purpose of hash key/partition key and how they are managed by DynamoDB vs ScyllaDB

tzach commented on docs/alternator/compatibility.md:
```diff
@@ -16,7 +16,21 @@ provision that - Scylla requires you to provision your cluster. You need
+
+In DynamoDB, the Hash key (or partition key) determines where the item will
+be stored within DynamoDB's internal storage. Another notable difference between
+DynamoDB and Scylla comes down to the underlying hashing algorithm.
```
```suggestion
DynamoDB and Scylla come down to the underlying hashing algorithm.
```

nyh replied:
@tzach, "comes" was actually correct - "another notable difference ... comes".

tzach replied:
I got this input from Grammerly, but I trust you more.

nyh commented on docs/alternator/compatibility.md:
```diff
@@ -16,7 +16,21 @@ provision that - Scylla requires you to provision your cluster. You need
+be stored within DynamoDB's internal storage. Another notable difference between
+DynamoDB and Scylla comes down to the underlying hashing algorithm.
+While DynamoDB uses a proprietary hashing function, ScyllaDB implements the well-known
+[Murmur3](https://docs.scylladb.com/stable/glossary.html#term-Partitioner) algorithm.
```
I think there's a parallel patch to add a separate "mumur3" entry to the glossary.
By the way, unfortunately Cassandra failed to implement the "well-known" murmur3 algorithm (it had a bug in the sign bit...), and Scylla (deliberately) copied its buggy implementation. But I guess this is beyond what needs to appear in this page.

tzach replied:
https://github.com/scylladb/scylladb/pull/11369
but not merge yet, and not a must here
